### PR TITLE
Bookmarks

### LINKED ISSUE #511: Add a `moz_bookmarks` table
This will use mostly the same schema as Desktop. Talking with @mhammond on Slack, we decided to store the `parent` ID and `position` in `moz_bookmarks`, instead of storing the structure in a separate table. If we want to use a closure table, we can always migrate later. Desktop's schema is well-understood and tested, and two-pass application for synced bookmarks (first pass to insert new items, second pass to update the structure) isn't that expensive.

Keywords can be a property of the bookmark for round-tripping; we can ignore `moz_keywords`.

### LINKED ISSUE #512: Port `moz_places.foreign_count` triggers from Desktop
We'll need to bump these for bookmarks and tags, and, eventually, favicons.

### LINKED ISSUE #514: Add a `moz_bookmarks_deleted` table for tombstones
Let's copy Desktop to start with: delete tombstones after upload, only write tombstones for `NORMAL` (not `NEW` or `UNKNOWN`) items. Let's also add a trigger to ensure a bookmark can't exist in both `moz_bookmarks` and `moz_bookmarks_deleted`; we do this manually now.

### LINKED ISSUE #516: Build a small API that exercises the bookmarks schema
* Add, update, and delete bookmarks.
* Sync-aware (writes tombstones, etc.), but doesn't need to sync yet.

### PR DESCRIPTION
**Pull request by mhammond**
This is intended to be a "long lived" PR to track the work in the "bookmarks" branch. This branch is where we can collaborate on bookmarks until we get to a point where we consider it "done enough" to flatten and merge into master.

Closes #511. Closes #512. Closes #514. Closes #516.

### CODE REVIEW

File: components/places/src/storage/bookmarks.rs [hunk 1]
```diff
@@ -0,0 +1,1102 @@
...
+                            log::warn!("ignoring malformed bookmark {:?}- no URL", row);
+                            None
+                        }
+                    },
+                    BookmarkType::Separator => Some(BookmarkTreeNode::Separator(SeparatorNode {
+                        guid: Some(row.guid.clone()),
+                        date_added: Some(row.date_added),
+                        last_modified: Some(row.last_modified),
+                    })),
+                    BookmarkType::Folder => panic!("impossible - we already peeked and checked"),
```
Comment (linacambridge) [inline components/places/src/storage/bookmarks.rs]: Nit: `unreachable!("...")`.

File: components/places/examples/places_utils.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+}
+
+fn convert_timestamp(t: Option<u64>) -> Option<Timestamp> {
+    match t {
+        None => None,
+        Some(v) => Some(Timestamp(v / 1000)),
+    }
+}
+
+fn convert_node(dm: DesktopMapping) -> Option<BookmarkTreeNode> {
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/examples/places_utils.rs]: I'm not convinced that we should implement this in the library itself (ie, I don't think it should be a goal that the library can import/export to/from bookmarks), so I didn't want to add it to bookmarks.rs, and I can't add it to BookmarkTreeNode in this file as it's not considered as being in the same crate.

File: components/places/examples/places_utils.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+            Some(_) => BookmarkType::Bookmark,
+            _ => BookmarkType::Folder,
+        },
+    };
+    Some(match bookmark_type {
+        BookmarkType::Bookmark => {
+            let url = match dm.uri {
+                Some(uri) => uri,
+                None => {
+                    log::warn!("ignoring bookmark node without url: {:?}", dm);
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/examples/places_utils.rs]: This is recursive so we are the caller :) It's also just a little utility and not part of the library, so I don't think carefully managing logging is that important - so I left it alone, but feel free to push back if you feel strongly.

File: components/places/examples/places_utils.rs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+fn do_import(db: &PlacesDb, root: BookmarkTreeNode) -> Result<()> {
+    // We need to import each of the sub-trees individually.
+    // Later we will want to get smarter around guids - currently we will
+    // fail to do this twice due to guid dupes - but that's OK for now.
+    let folder = match root {
+        BookmarkTreeNode::Folder(folder_node) => folder_node,
+        _ => {
+            println!("Imported node isn't a folder structure");
+            return Ok(());
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/examples/places_utils.rs]: As above, I don't care too much about that - the tool prints an error and exits. I guess a non-zero exit code makes sense, but at this stage I'm not sure it's worth it.

File: components/places/src/db/db.rs [hunk 5]
```diff
@@ -68,6 +69,9 @@ impl PlacesDb {
             -- UnifiedComplete). Note that SQLite uses a negative value for this pragma to indicate
             -- that it's in units of KiB.
             PRAGMA cache_size = -6144;
+
+            -- We want foreign-key support.
+            PRAGMA foreign_keys = ON;
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/src/db/db.rs]: Thanks!

File: components/places/src/db/schema.rs [hunk 6]
```diff
@@ -163,6 +179,72 @@ lazy_static! {
...
+        SET foreign_count = foreign_count - 1
+        WHERE id = OLD.fk;
+    END";
+
+const CREATE_TRIGGER_BOOKMARKS_FOREIGNCOUNT_AFTERINSERT_TRIGGER: &str = "
+    CREATE TEMP TRIGGER moz_bookmarks_foreign_count_afterinsert_trigger
+    AFTER INSERT ON moz_bookmarks FOR EACH ROW
+    BEGIN
+-- hrm        SELECT store_last_inserted_id('moz_bookmarks', NEW.id);
+-- hrm        SELECT note_sync_change() WHERE NEW.syncChangeCounter > 0;
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/src/db/schema.rs]: Thanks! I removed the "hrm" comments and replaced it with a more descriptive comment above the defn.

File: components/places/src/db/schema.rs [hunk 7]
```diff
@@ -163,6 +179,72 @@ lazy_static! {
...
+-- hrm      WHERE NEW.syncChangeCounter <> OLD.syncChangeCounter;
+        UPDATE moz_places
+            SET foreign_count = foreign_count + 1
+            WHERE OLD.fk <> NEW.fk AND id = NEW.fk;
+        UPDATE moz_places
+            SET foreign_count = foreign_count - 1
+            WHERE OLD.fk <> NEW.fk AND id = OLD.fk;
+    END";
+
+// Unlike history, we manage bookmark tombstones via triggers. We do this
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/src/db/schema.rs]: Yeah, although we don't really have that concept yet (but we will need it soon!)

File: components/places/src/db/schema.rs [hunk 8]
```diff
@@ -163,6 +179,72 @@ lazy_static! {
...
+
+// and also for updating the guid.
+const CREATE_TRIGGER_BOOKMARKS_REMOVE_TOMBSTONE_UPDATE: &str = "
+    CREATE TEMP TRIGGER moz_remove_bookmarks_deleted_update_trigger
+    AFTER UPDATE ON moz_bookmarks
+    FOR EACH ROW WHEN OLD.guid != NEW.guid
+    BEGIN
+        DELETE from moz_bookmarks_deleted WHERE guid = NEW.guid;
+    END";
+// XXX - we should possibly insert a bookmark for OLD.guid here, although
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/src/db/schema.rs]: good idea! I also added a comment about the fact we expect sync connections to not have these triggers.

File: components/places/src/db/schema.rs [hunk 9]
```diff
@@ -74,20 +76,34 @@ const CREATE_TABLE_INPUTHISTORY_SQL: &str = "CREATE TABLE moz_inputhistory (
...
-// test autocomplete.
-const CREATE_TABLE_BOOKMARKS_SQL: &str = "CREATE TABLE moz_bookmarks (
+const CREATE_TABLE_BOOKMARKS_SQL: &str = r#"CREATE TABLE moz_bookmarks (
         id INTEGER PRIMARY KEY,
-        fk INTEGER,
-        title TEXT,
+        fk INTEGER DEFAULT NULL, -- place_id
+        type INTEGER NOT NULL CHECK(type BETWEEN 1 AND 3),
+        parent INTEGER,
+        position INTEGER NOT NULL CHECK(position >= 0),
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/src/db/schema.rs]: Interesting :) I've done nothing here for now.

File: components/places/src/db/schema.rs [hunk 10]
```diff
@@ -163,6 +179,72 @@ lazy_static! {
...
+    FOR EACH ROW WHEN OLD.syncStatus = 2 -- SyncStatus::Normal
+    BEGIN
+        INSERT into moz_bookmarks_deleted VALUES (OLD.guid, now());
+    END";
+
+// And a trigger to ensure adding a previously deleted bookmark with the same
+// guid doesn't leave an entry in both tables.
+const CREATE_TRIGGER_BOOKMARKS_REMOVE_TOMBSTONE_INSERT: &str = "
+    CREATE TEMP TRIGGER moz_remove_bookmarks_deleted_insert_trigger
+    AFTER INSERT ON moz_bookmarks
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/src/db/schema.rs]: Updated this trigger, the update trigger below ends up RAISE(FAIL, ...)ing.

File: components/places/src/storage/bookmarks.rs [hunk 11]
```diff
@@ -0,0 +1,1102 @@
...
+
+fn create_root(
+    db: &Connection,
+    title: &str,
+    guid: &SyncGuid,
+    position: u32,
+    when: &Timestamp,
+) -> Result<()> {
+    // desktop's sql seems to assume the root gets a rowid of zero, whereas
+    // we see 1 here. Regardless, the sql below uses the guid.
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/src/storage/bookmarks.rs]: https://searchfox.org/mozilla-central/rev/6c784c93cfbd5119ed07773a170b59fbce1377ea/toolkit/components/places/Database.cpp#240 - but hrmph - `root________` is the actual root here.

File: components/places/src/storage/bookmarks.rs [hunk 12]
```diff
@@ -0,0 +1,1102 @@
...
+    use crate::db::PlacesDb;
+
+    #[test]
+    fn test_insert() -> Result<()> {
+        let _ = env_logger::try_init();
+        let conn = PlacesDb::open_in_memory(None)?;
+        let url = Url::parse("https://www.example.com")?;
+
+        let bm = InsertableBookmarkItem::Bookmark(InsertableBookmark {
+            parent_guid: BookmarkRootGuid::Unfiled.as_guid(),
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/src/storage/bookmarks.rs]: Yep - not "everywhere" - eg, the asserts still require as_guid, but I made the changes where I could.

File: components/places/src/storage/bookmarks.rs [hunk 13]
```diff
@@ -0,0 +1,1102 @@
...
+    }
+    impl_common_bookmark_getter!(parent_guid, SyncGuid);
+    impl_common_bookmark_getter!(position, BookmarkPosition);
+    impl_common_bookmark_getter!(date_added, Option<Timestamp>);
+    impl_common_bookmark_getter!(last_modified, Option<Timestamp>);
+    impl_common_bookmark_getter!(guid, Option<SyncGuid>);
+}
+
+pub fn insert_bookmark(db: &impl ConnExt, bm: InsertableBookmarkItem) -> Result<SyncGuid> {
+    let tx = db.unchecked_transaction()?;
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/src/storage/bookmarks.rs]: If you hate this name, wait until you see #577 ;)

File: components/places/src/storage/bookmarks.rs [hunk 14]
```diff
@@ -0,0 +1,1102 @@
...
+
+impl<'de> Deserialize<'de> for BookmarkTreeNode {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // *sob* - a union of fields we post-process.
+        #[derive(Debug, Default, Deserialize)]
+        #[serde(default)]
+        struct Mapping {
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/places/src/storage/bookmarks.rs]: As above, DesktopMapping isn't part of the lib - this is the "official" ;) json-import-export support by the lib and used extensively by the tests - we should consider changes here to make it more consistent with desktop, but I don't think compatibility with desktop needs to be a goal of the lib. Feel free to push back though!

File: components/places/examples/places_utils.rs [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+fn init_logging() {
+    // Explicitly ignore some rather noisy crates. Turn on trace for everyone else.
+    let spec = "trace,tokio_threadpool=warn,tokio_reactor=warn,tokio_core=warn,tokio=warn,hyper=warn,want=warn,mio=warn,reqwest=warn";
+    env_logger::init_from_env(env_logger::Env::default().filter_or("RUST_LOG", spec));
+}
+
+// A struct in the format of desktop with a union of all fields.
+#[derive(Debug, Default, Deserialize)]
+#[serde(default)]
+#[allow(non_snake_case)]
```
Comment (thomcc) [inline components/places/examples/places_utils.rs]: I'd prefer we use serde to rename non snake case items.

File: components/places/examples/places_utils.rs [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+use sql_support::ConnExt;
+use std::fs::File;
+use std::io::{BufReader, BufWriter};
+use url::Url;
+
+type Result<T> = std::result::Result<T, failure::Error>;
+
+fn init_logging() {
+    // Explicitly ignore some rather noisy crates. Turn on trace for everyone else.
+    let spec = "trace,tokio_threadpool=warn,tokio_reactor=warn,tokio_core=warn,tokio=warn,hyper=warn,want=warn,mio=warn,reqwest=warn";
```
Comment (thomcc) [inline components/places/examples/places_utils.rs]: God we need to put this somewhere shared. Some support/examples or something. (it's fine for now, just a thought).

File: components/places/examples/places_utils.rs [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+
+    let tree = fetch_tree(db.conn(), &BookmarkRootGuid::Root.into())?.unwrap();
+    serde_json::to_writer_pretty(writer, &tree)?;
+    Ok(())
+}
+
+fn main() -> Result<()> {
+    init_logging();
+
+    let matches = clap::App::new("places_utils")
```
Comment (thomcc) [inline components/places/examples/places_utils.rs]: In the future we should start using structopt (which generates the `clap` calls), as in https://github.com/mozilla/application-services/blob/master/testing/sync-test/src/main.rs#L54-L90. But this is fine unless you feel inspired to change it, just a note.

File: components/places/src/storage/bookmarks.rs [hunk 18]
```diff
@@ -0,0 +1,1101 @@
...
+            #[serde(rename = "type")]
+            bookmark_type: u8,
+            guid: Option<SyncGuid>,
+            date_added: Option<Timestamp>,
+            last_modified: Option<Timestamp>,
+            title: Option<String>,
+            #[serde(with = "url_serde")]
+            url: Option<Url>,
```
Comment (thomcc) [inline components/places/src/storage/bookmarks.rs]: We had issues with using url_serde here in https://github.com/mozilla/application-services/issues/571 :(

File: components/places/src/db/schema.rs [hunk 19]
```diff
@@ -74,20 +76,34 @@ const CREATE_TABLE_INPUTHISTORY_SQL: &str = "CREATE TABLE moz_inputhistory (
...
+        fk INTEGER DEFAULT NULL, -- place_id
+        type INTEGER NOT NULL CHECK(type BETWEEN 1 AND 3),
+        parent INTEGER,
+        position INTEGER NOT NULL CHECK(position >= 0),
+        title TEXT, -- a'la bug 1356159, NULL is special here - it means 'not edited'
+        dateAdded INTEGER NOT NULL DEFAULT 0,
         lastModified INTEGER NOT NULL DEFAULT 0,
+        guid TEXT NOT NULL UNIQUE CHECK(length(guid) == 12),
+
+        syncStatus INTEGER NOT NULL DEFAULT 0 CHECK(syncStatus BETWEEN 0 AND 2),
```
Comment (thomcc) [inline components/places/src/db/schema.rs]: How easy will it be for us to migrate to allow sync status == 3 here? (for Unknown)? is there a way to to change column constraints in alter table?

Comment (mhammond) [reply components/places/src/db/schema.rs]: > How easy will it be for us to migrate to allow sync status == 3 here? (for Unknown)?

SyncStatus::Unknown is zero.

> is there a way to to change column constraints in alter table?

Apparently not - iiuc, you need to create a new "temp" table with the correct constraints when select the entire old table into it.

File: components/places/src/db/schema.rs [hunk 20]
```diff
@@ -362,4 +455,374 @@ mod tests {
...
+        let e = conn
+            .execute_cached(
+                "INSERT INTO moz_bookmarks
+                (fk, type, parent, position, dateAdded, lastModified, guid)
+            VALUES
+                (NULL, 4, 0, 0, 1, 1, 'fake_guid___')",
+                NO_PARAMS,
+            )
+            .expect_err("should fail");
+        // it's a shame the message doesn't indicate what check failed :(
```
[the remark this answers is not in the document]
Comment (thomcc) [reply components/places/src/db/schema.rs]: `#[cfg(feature = "bundled")]` is used for a bunch of things that require newer versions of sqlite than are common. It's not ideal.

File: components/places/src/db/schema.rs (hunk 19 above)
Comment (thomcc) [reply components/places/src/db/schema.rs]: After discussions, we've removed some of these.

File: components/places/src/db/schema.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (thomcc) [reply components/places/src/db/schema.rs]: Since we just want this for tests its probably fine.

File: components/places/src/types.rs [hunk 21]
```diff
@@ -214,6 +214,21 @@ impl BookmarkType {
             _ => None,
         }
     }
+
+    pub fn from_u8_with_valid_url<F: Fn() -> bool>(v: u8, has_valid_url: F) -> Self {
```
Comment (linacambridge) [inline components/places/src/types.rs]: I think this needs a better name, but I'm not sure what. 😅

File: components/places/src/storage/bookmarks.rs [hunk 22]
```diff
@@ -883,21 +814,79 @@ pub fn fetch_tree(db: &impl ConnExt, item_guid: &SyncGuid) -> Result<Option<Book
...
+                last_modified: Some(row.last_modified),
+                title: row.title.clone(),
+                children: Vec::new(),
+            }
+            .into()
+        }
+        None => return Ok(None),
+    };
+
+    // For all remaining rows, build a pseudo-tree that maps parent GUIDs to
```
Comment (linacambridge) [inline components/places/src/storage/bookmarks.rs]: @thomcc, could you double-check 👀 this, please?

File: components/places/src/storage/bookmarks.rs [hunk 23]
```diff
@@ -883,21 +814,79 @@ pub fn fetch_tree(db: &impl ConnExt, item_guid: &SyncGuid) -> Result<Option<Book
...
+                        date_added: Some(row.date_added),
+                        last_modified: Some(row.last_modified),
+                        title: row.title.clone(),
+                        url,
+                    }
+                    .into(),
+                    Err(e) => {
+                        log::warn!(
+                            "ignoring malformed bookmark - invalid URL {}: {:?}",
+                            url_str,
```
Comment (thomcc) [inline components/places/src/storage/bookmarks.rs]: PII in log::warn! isn't good.

File: components/places/src/storage/bookmarks.rs (hunk 22 above)
Comment (thomcc) [reply components/places/src/storage/bookmarks.rs]: `HashMap<Option<SyncGuid>, Vec<BookmarkTreeNode>>` huh that's a bit weird. Guess it makes some sense though.

File: components/places/src/storage/bookmarks.rs [hunk 24]
```diff
@@ -883,21 +814,79 @@ pub fn fetch_tree(db: &impl ConnExt, item_guid: &SyncGuid) -> Result<Option<Book
...
+                            e
+                        );
+                        continue;
+                    }
+                },
+                None => {
+                    log::warn!("ignoring malformed bookmark {:?}- no URL", row);
```
Comment (thomcc) [inline components/places/src/storage/bookmarks.rs]: Ditto.

**Pull request merged**